### PR TITLE
[SPARK-31489][SQL] Fix pushing down filters with `java.time.LocalDate` values in ORC

### PR DESCRIPTION
**Pull request by MaxGekk**
### What changes were proposed in this pull request?
Convert `java.time.LocalDate` to `java.sql.Date` in pushed down filters to ORC datasource when Java 8 time API enabled.

Closes #28272

### Why are the changes needed?
The changes fix the exception raised while pushing date filters when `spark.sql.datetime.java8API.enabled` is set to `true`:
```
Wrong value class java.time.LocalDate for DATE.EQUALS leaf
java.lang.IllegalArgumentException: Wrong value class java.time.LocalDate for DATE.EQUALS leaf
	at org.apache.hadoop.hive.ql.io.sarg.SearchArgumentImpl$PredicateLeafImpl.checkLiteralType(SearchArgumentImpl.java:192)
	at org.apache.hadoop.hive.ql.io.sarg.SearchArgumentImpl$PredicateLeafImpl.<init>(SearchArgumentImpl.java:75)
	at org.apache.hadoop.hive.ql.io.sarg.SearchArgumentImpl$BuilderImpl.equals(SearchArgumentImpl.java:352)
	at org.apache.spark.sql.execution.datasources.orc.OrcFilters$.buildLeafSearchArgument(OrcFilters.scala:229)
```

### Does this PR introduce any user-facing change?
Yes

### How was this patch tested?
Added tests to `OrcFilterSuite`.

### CODE REVIEW

**Comment by MaxGekk** [comment]
jenkins, retest this, please

**Comment by MaxGekk** [comment]
@dongjoon-hyun @cloud-fan @HyukjinKwon Could you take a look at the PR, please.

File: sql/core/v2.3/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcFilterSuite.scala [hunk 1]
```diff
@@ -451,5 +452,31 @@ class OrcFilterSuite extends OrcTest with SharedSparkSession {
       ).get.toString
     }
   }
+
+  test("filter pushdown - local date") {
+    val dates = Seq("2017-08-18", "2017-08-19", "2017-08-20", "2017-08-21").map { day =>
+      LocalDate.parse(day)
+    }
+    withSQLConf(SQLConf.DATETIME_JAVA8API_ENABLED.key -> "true") {
```
Comment (cloud-fan) [inline sql/core/v2.3/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcFilterSuite.scala:460]: It makes me realize that this config can break data source v1/v2 as it changes the value in `Filter`.

Shall we fix `DataSourceAnalysis.translateLeafNodeFilter` instead to always return java 7 Date/Timestamp?

Comment (MaxGekk) [reply sql/core/v2.3/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcFilterSuite.scala:460]: @cloud-fan Here is the PR https://github.com/apache/spark/pull/28272

**Comment by dongjoon-hyun** [comment]
Thank you for pinging me, @MaxGekk .
Could you handle the following together?
- sql/core/v1.2/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcFilters.scala

**Comment by dongjoon-hyun** [comment]
Retest this please.

File: sql/core/v2.3/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcFilterSuite.scala [hunk 2]
```diff
@@ -451,5 +452,31 @@ class OrcFilterSuite extends OrcTest with SharedSparkSession {
       ).get.toString
     }
   }
+
+  test("filter pushdown - local date") {
```
Comment (dongjoon-hyun) [inline sql/core/v2.3/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcFilterSuite.scala:456]: nit. `test("SPARK-31489: filter`.

**Comment by dongjoon-hyun** [comment]
For this PR, please pass the PR builders with both Hive 1.2 and Hive 2.3.

File: sql/core/v2.3/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcFilterSuite.scala (hunk 1 above)
Comment (MaxGekk) [reply sql/core/v2.3/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcFilterSuite.scala:460]: @dongjoon-hyun ^^ The PR covers this PR

Comment (dongjoon-hyun) [reply sql/core/v2.3/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcFilterSuite.scala:460]: Got it. Thanks, @MaxGekk !

**Comment by MaxGekk** [comment]
jenkins, retest this, please

**Comment by cloud-fan** [comment]
retest this please

**Comment by MaxGekk** [comment]
jenkins, retest this, please

**Comment by MaxGekk** [comment]
jenkins, retest this, please

**Comment by MaxGekk** [comment]
jenkins, retest this, please

**Comment by MaxGekk** [comment]
jenkins, retest this, please

**Comment by MaxGekk** [comment]
jenkins, retest this, please

**Comment by MaxGekk** [comment]
This PR has been tested for hive-1.2 and hive-2.3, and it is ready for review. @cloud-fan @dongjoon-hyun Please, take a look at it.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you, @MaxGekk and @cloud-fan .
Merged to master/3.0.

**Pull request closed without merging**